### PR TITLE
Optimizing performance for destroying Study objects (SCP-2940)

### PR DESCRIPTION
**Pull request by bistline**
Due to how Rails handles callbacks on associations when invoking the `destroy` method, doing this for a `Study` object has severe performance implications both in local development and on deployed hosts.  Calling `study.destroy` can essentially grind performance to a halt as the ORM performs expensive collections walks collections with potentially millions of documents, and then calls `destroy` or `delete` on individual associated documents.  In production, this can take over an hour, during which time the portal may become unstable.  Normal study deletions from the UI/API do not incur this penalty, as they rely on the `DeleteQueueJob` class to manage deletes in a background process that runs on a nightly basis, after the database has been reindexed to optimize performance.

This update now removes all `dependent: :destroy` or `dependent: :delete` association callbacks in the `Study` class on the largest collections containing "parsed" data, such as `CellMetadatum`, `Gene`, `ClusterGroup`, and `DataArray`.  A new `before_destroy` callback now ensures that all previously dependent associations preemptively delete data in a performant fashion, and then the parent `Study` object is destroyed, which will remove all other dependent records.  In addition, calling `study.destroy` will now delete convention metadata from BigQuery, without the need to run `DeleteQueueJob`.

Note: this update does not change how `DeleteQueueJob` is invoked or managed - this only applies to calling `study.destroy` directly.   

This PR satisfies SCP-2940.

### CODE REVIEW

File: app/models/study.rb [hunk 1]
```diff
@@ -1635,6 +1622,18 @@ def self.delete_all_and_remove_workspaces
     end
   end
 
+  # helper method that mimics DeleteQueueJob.delete_convention_data
+  # referenced from ensure_cascade_on_associations to prevent orphaned rows in BQ on manual deletes
```
Comment (devonbush) [inline app/models/study.rb:1626]: I like this method a lot, should we update DeleteQueueJob to call this method instead of its own?

**Review by devonbush — APPROVED** [review]
looks great, appreciate the deep dive you took here.  Just one refactor question

File: app/models/study.rb (hunk 1 above)
Comment (bistline) [reply app/models/study.rb:1626]: Originally I thought "no", but on reflection maybe it should.  I'll check it out.

File: app/models/study.rb [hunk 2]
```diff
@@ -1959,4 +1959,26 @@ def prevent_firecloud_attribute_changes
       end
     end
   end
+
+  # delete all records that are associate with this study before invoking :destroy to speed up performance
```
Comment (eweitz) [inline app/models/study.rb:1963]:
```suggestion
  # delete all records that are associated with this study before invoking :destroy to speed up performance
```

File: test/integration/download_agreement_test.rb [hunk 3]
```diff
@@ -21,8 +20,7 @@ class DownloadAgreementTest < ActionDispatch::IntegrationTest
   end
 
   teardown do
-    @study = Study.find_by(name: "Download Agreement #{@random_seed}")
-    delete_study_and_ensure_cascade(@study)
+    Study.find_by(name: "Download Agreement #{@random_seed}").destroy
```
Comment (eweitz) [inline test/integration/download_agreement_test.rb:23]: More succinct _and_ faster?  Cool.

**Review by eweitz — APPROVED** [review]

**Pull request merged**